### PR TITLE
Cabolanos.90.riak backend

### PR DESCRIPTION
**Pull request by cabol**
First version of Riak backend implementation finished.
 * Riak Data Types as main structures to push/pull data.
 * Bulk operations (such as: delete_all and find_all) were optimized using streaming. Records are streamed in portions (using Riak 2i to stream keys first), and then the current operation (e.g.: delete the record or accumulate the values to return them later) is applied. This allows better memory and cpu efficiency.
 * Query functions were implemented using Riak Search on Data Types, to get better performance and flexibility.
* Multi-level support for basic K/V operations (persist, find_all, delete_all, find_by ID, delete_by ID). TODO: add multilevel support to search operations (basically: delete_by/find_by query options).

### CODE REVIEW

File: src/sumo_store_riak.erl [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    {_, Count}  -> {error, Count, State}
+  end.
+
+-spec find_all(
+  sumo:schema_name(), state()
+) -> sumo_store:result([sumo_internal:doc()], state()).
+find_all(DocName, #state{conn = Conn, bucket = Bucket} = State) ->
+  Get = fun({C, B, Kst}, Acc) ->
+          Fun = fun(K, Acc0) ->
+                  case fetch_map(C, B, K) of
```
Comment (elvisci) [inline src/sumo_store_riak.erl]: The expression on line 128 and column 19 is nested beyond the maximum level of 3.

File: src/sumo_store_riak.erl [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+  map_to_rmap(Fields).
+
+%% @private
+%% Support multi-level structures.
+map_to_rmap(Map) ->
+  F = fun({K, V}, Acc) ->
+        case V of
+          V when is_map(V) ->
+            NewV = map_to_rmap(V),
+            riakc_map:update({to_bin(K), map}, fun(_M) -> NewV end, Acc);
```
Comment (elvisci) [inline src/sumo_store_riak.erl]: The expression on line 241 and column 48 is nested beyond the maximum level of 3.

File: src/sumo_store_riak.erl [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+          V when is_list(V) ->
+            case io_lib:printable_list(V) of
```
Comment (elvisci) [inline src/sumo_store_riak.erl]: The expression on line 243 and column 13 is nested beyond the maximum level of 3.

File: src/sumo_store_riak.erl [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+                  {to_bin(K), register},
+                  fun(R) -> riakc_register:set(to_bin(V), R) end,
+                  Acc);
+              false ->
+                riakc_map:update({to_bin(K), set}, fun(_S) -> V end, Acc)
+            end;
+          _ ->
+            riakc_map:update(
+              {to_bin(K), register},
+              fun(R) -> riakc_register:set(to_bin(V), R) end,
```
Comment (elvisci) [inline src/sumo_store_riak.erl]: The expression on line 255 and column 15 is nested beyond the maximum level of 3.

File: test/sumo_basic_SUITE.erl [hunk 5]
```diff
@@ -87,20 +87,38 @@ init_store(Module) ->
...
-  %% This is necessary to get elasticsearch in
-  %% particular to index its stuff.
-  timer:sleep(1000).
+  %% Sync Timeout.
+  %% 1. Necessary to get elasticsearch in particular to index its stuff.
+  %% 2. Necessary to Riak.
+  %% Riak clusters will retain deleted objects for some period of time
+  %% (3 seconds by default), and the MapReduce framework does not conceal
+  %% these from submitted jobs.
+  %% @see <a href="http://docs.basho.com/riak/latest/dev/advanced/mapreduce"></a>
```
Comment (elvisci) [inline test/sumo_basic_SUITE.erl]: Line 96 is too long:   %% @see <a href="http://docs.basho.com/riak/latest/dev/advanced/mapreduce"></a>.

File: README.md [hunk 6]
```diff
@@ -71,6 +71,76 @@ being in the top level directory:
...
+    $ curl -XPUT $RIAK_HOST/search/index/sumo_test_index \
+        -H 'Content-Type: application/json' \
+        -d '{"schema":"_yz_default"}'
+
+With our index created, we can associate our new `sumo_test_index` index with
+our `maps` bucket type:
+
+    $ riak-admin bucket-type update maps '{"props":{"search_index":"sumo_test_index"}}'
+
+Now we can start to working with **Riak** from `sumo_db`.
```
Comment (elbrujohalcon) [inline README.md]: …start working…

File: README.md [hunk 7]
```diff
@@ -71,6 +71,76 @@ being in the top level directory:
...
+Due to **Riak** comes with default configuration, we need to change some
+parameters required by `sumo_db`.
+
+**Riak** has a main configuration file `riak.conf`, which you can find into
+your installation path `$YOUR_INSTALL_PATH/etc/riak.conf`.
+
+> **Note:** For more information check this link [Configuration Files](http://docs.basho.com/riak/latest/ops/advanced/configs/configuration-files).
+
+First parameter to change is the default **Riak** backend from **Bitcask** to
+**LevelDB**. This change also eblables to use [Riak Secondary Indexes](http://docs.basho.com/riak/latest/ops/advanced/configs/secondary-index/).
```
Comment (elbrujohalcon) [inline README.md]: …eNables the use of… instead of …eblables to use…

File: README.md [hunk 8]
```diff
@@ -71,6 +71,76 @@ being in the top level directory:
...
+ * For **MySQL**, **PostgreSQL** and **MongoDB**, you need to:
+    - Create an user (or use defaults) and configure it on `test/test.config`
+      file.
+    - Create test database `sumo_test` on each DB.
+
+> **Note:**
+
+> - For **MongoDB** you first create the test database and then create an user
+    to access that DB. For more information visit [MongoDB Tutorial](http://docs.mongodb.org/manual/tutorial).
+> - For **Riak** please follow instruction below first ([<i class="icon-refresh"></i> Riak](#riak)).
```
Comment (elbrujohalcon) [inline README.md]: `- For **Riak** please follow instructions below.`

File: README.md [hunk 9]
```diff
@@ -71,6 +71,76 @@ being in the top level directory:
...
+## Riak
+
+### Install Riak
+
+To install/upgrade **Riak** please follow the instructions in this link:
+[Installing and Upgrading Riak](http://docs.basho.com/riak/latest/ops/building/installing).
+
+### Initial Configurations
+
+Due to **Riak** comes with default configuration, we need to change some
```
Comment (elbrujohalcon) [inline README.md]: Due to **the fact** that **Riak** …

File: src/sumo_backend_riak.erl [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  Port = proplists:get_value(port, Options, 8087),
+  Opts = riak_opts(Options),
+  BucketType = iolist_to_binary(
+    proplists:get_value(bucket_type, Options)),
+  Bucket = iolist_to_binary(
+    proplists:get_value(bucket, Options, <<"sumo_test">>)),
+  Index = iolist_to_binary(
+    proplists:get_value(index, Options, <<"sumo_test_index">>)),
+  %% Encoding
+  %%Encoding = proplists:get_value(encoding, Options, json),
```
Comment (elbrujohalcon) [inline src/sumo_backend_riak.erl]: Don't leave commented out code on git. Please remove it.

**Pull request merged**